Add quote ASTs to TASTy

nicolasstucki opened the pull request:
Add AST nodes for `Quote`, `Splice`, `QuotePattern`, and `QuoteSplice` to TASTy.


```
Term          = 
                  ...
                  QUOTE          Length body_Term bodyTpe_Type                     -- Quoted expression `'{ body }` of a body typed as `bodyTpe`
                  SPLICE         Length expr_Term tpe_Type                         -- Spliced expression `${ expr }` typed as `tpe`
                  SPLICEPATTEN   Length pat_Term tpe_Type targs_Type* args_Term*   -- Pattern splice `${pat}` or `$pat[targs*](args*)` in a quoted pattern of type `tpe`.
-- patterns:
                  ...
                  QUOTEPATTERN   Length body_Term quotes_Term pat_Type bindings_Term* -- Quote pattern node `'{ bindings*; body }(using quotes)`
```

nicolasstucki commented:
@WojciechMazur could you test this PR on the open community build?

WojciechMazur commented:
The OpenCB was started, I'll post the result tomorrow when it's done

WojciechMazur commented:
We have 4 projects that started to fail after this change. In most cases it seems like the typer issue, eg. 
```scala
Error:     |           Found:    Int => String
Error:     |           Required: T => Any
```
| Project | Version | Build URL | Notes |
| ------- | ------- | --------- | ----- |
| dwickern/scala-nameof | 4.0.0 | [Open CB logs](https://github.com/VirtusLab/community-build3/actions/runs/8737538643/job/23980574344) |  Type Mismatch Error for macro inputs |
| yakivy/jam | 0.4.5 | [Open CB logs](https://github.com/VirtusLab/community-build3/actions/runs/8737538643/job/23981797303) | Type Mismatch Error for macro inputs |
| zio/zio-direct | 1.0.0-RC7 | [Open CB logs](https://github.com/VirtusLab/community-build3/actions/runs/8737538643/job/23982293929) | Type Mismatch Error for macro inputs |
| zio/zio-prelude | 1.0.0-RC23 | [Open CB logs](https://github.com/VirtusLab/community-build3/actions/runs/8737538643/job/23981801720) | Type Mismatch Error for macro inputs |

nicolasstucki commented on compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:
```diff
@@ -701,28 +701,33 @@ class TreePickler(pickler: TastyPickler, attributes: Attributes) {
+            pickleTree(pat)
+            pickleType(tree.tpe)
+            args.foreach(pickleTree)
+            // TODO: Handle `targs` for #18271. We should be able to pickle the
+            //       `targs ::: args` if we add an `EXPLICITtpt` to the targs.
+            //       See comment in SPLICEPATTERN unpickling.
```
I previously thought that we could do it in a backward-compatible way, but now I see that my strategy will not work. I will propose another solution that will be future-proof for #18271.

nicolasstucki replied:
Actually my original idea will work. I replaced the comments with a partial implementation.

nicolasstucki commented:
> We have 4 projects that started to fail after this change. In most cases it seems like the typer issue, eg.
> ...

I fixed these with the change in 3f2db240cd71b0c8b48e3d3df6c6082821d8ac09. This also removed the changes in the checkfiles (these where suspicious).

Gedochao commented:
This has been decided to be included in the 3.5.0 release.
